network/src/noise/handshake.rs

```diff
@@ -523,4 +529,46 @@ mod test {
     fn test_handshake_mutual_auth() {
         test_handshake_success(true /* is_mutual_auth */);
     }
+
+    /// provide a function that will return the same given value as a timestamp
+    fn bad_timestamp(value: u64) -> impl Fn() -> [u8; AntiReplayTimestamps::TIMESTAMP_SIZE] {
+        move || value.to_le_bytes()
+    }
+
+    #[test]
+    fn test_timestamp_replay() {
+        // 1. generate peers
+        let ((client, _client_public), (server, server_public)) =
+            build_peers(true /* is_mutual_auth */);
+
+        // 2. perform the handshake with some timestamp, it should work
+        let (dialer_socket, listener_socket) = MemorySocket::new_pair();
+        let (client_session, server_session) = block_on(join(
+            client.upgrade_outbound(dialer_socket, server_public, bad_timestamp(0)),
+            server.upgrade_inbound(listener_socket),
+        ));
+
+        client_session.unwrap();
+        server_session.unwrap();
+
+        // 3. perform the handshake again with the same timestamp, it should fail
+        let (dialer_socket, listener_socket) = MemorySocket::new_pair();
+        let (client_session, server_session) = block_on(join(
+            client.upgrade_outbound(dialer_socket, server_public, bad_timestamp(0)),
+            server.upgrade_inbound(listener_socket),
+        ));
```

Comment:
just need one more case: test that it also fails with a timestamp in the past, not just the same timestamp.

Reply:
good point!